core/src/main/java/tachyon/metrics/MetricsSystem.java

```diff
@@ -0,0 +1,206 @@
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.MetricRegistry;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.metrics.sink.Sink;
+import tachyon.metrics.source.Source;
+
+/**
+ * A MetricsSystem is created by a specific instance(master, worker). It polls the metrics sources
+ * periodically and pass the data to the sinks.
+ *
+ * The syntax of the metrics configuration file is:
+ * [instance].[sink|source].[name].[options]=[value]
+ */
+public class MetricsSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  public static final String SINK_REGEX = "^sink\\.(.+)\\.(.+)";
+  public static final String SOURCE_REGEX = "^source\\.(.+)\\.(.+)";
+  private static final TimeUnit MINIMAL_POLL_UNIT = TimeUnit.SECONDS;
+  private static final int MINIMAL_POLL_PERIOD = 1;
+
+  private String mInstance;
+  private List<Sink> mSinks = new ArrayList<Sink>();
+  private List<Source> mSources = new ArrayList<Source>();
+  private MetricRegistry mMetricRegistry = new MetricRegistry();
+  private MetricsConfig mMetricsConfig;
+  private boolean mRunning = false;
+  private TachyonConf mTachyonConf;
+
+  /**
+   * Check if the poll period is smaller that the minimal poll period which is 1 second.
+   *
+   * @param pollUnit the polling unit.
+   * @param pollPeriod the polling period.
+   * @throws IllegalArgumentException
+   */
+  public static void checkMinimalPollingPeriod(TimeUnit pollUnit, int pollPeriod)
+      throws IllegalArgumentException {
+    int period = (int) MINIMAL_POLL_UNIT.convert(pollPeriod, pollUnit);
+    if (period < MINIMAL_POLL_PERIOD) {
+      throw new IllegalArgumentException("Polling period " + pollPeriod + " " + pollUnit
+          + " is below than minimal polling period");
+    }
+  }
+
+  /**
+   * Creates a MetricsSystem and initialize the configuration.
+   *
+   * @param instance the instance name.
+   */
+  public MetricsSystem(String instance, TachyonConf tachyonConf) {
+    mInstance = instance;
+    mTachyonConf = tachyonConf;
+    mMetricsConfig = new MetricsConfig(mTachyonConf.get(Constants.METRICS_CONF_FILE, null));
+    mMetricsConfig.initialize();
+  }
+
+  /**
+   * Gets the sinks. Used by unit tests only.
+   *
+   * @return A list of registered Sinks.
+   */
+  public List<Sink> getSinks() {
+    return mSinks;
+  }
+
+  /**
+   * Gets the sources. Used by unit tests only.
+   *
+   * @return A list of registered Sources.
+   */
+  public List<Source> getSources() {
+    return mSources;
+  }
+
+  /**
+   * Register a Source.
+   *
+   * @param source the source to register.
+   */
+  public void registerSource(Source source) {
+    mSources.add(source);
+    try {
+      mMetricRegistry.register(source.getName(), source.getMetricRegistry());
+    } catch (IllegalArgumentException e) {
+      LOG.info("Metrics already registered", e);
```

Comment:
I changed this to WARN. The exception means the source has already been registered, so we don't need add it again.